### PR TITLE
Remove arraylet ifdefs, permanently enabling them

### PR DESCRIPTION
**Pull request by rwy0717**
This PR removes the arraylet ifdefs from J9, so that they are effectively always-on. In j9, building without arraylets hasn't been done in quite some time. In OMR, arraylets are already always-on. N

The flags which are permanently enabled:
- `OMR_GC_ARRAYETS`
- `OMR_GC_HYBRID_ARRAYLETS`
- `J9VM_GC_ARRAYLETS`
- `J9VM_GC_HYBRID_ARRAYLETS`

The flags are removed from the source, the build system, and ddr.

### CODE REVIEW

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java [hunk 1]
```diff
@@ -156,27 +146,20 @@ protected UDATA getSpineSizeWithoutHeader(long layout, UDATA numberArraylets, UD
 		UDATA spineArrayoidSize = new UDATA(0);
 		UDATA spinePaddingSize = new UDATA(0);
 
-		/* The spine consists of three (possibly empty) sections, not including the header:
+		/* The spine consists of three or four (possibly empty) sections, not including the header:
 		 * 1. the alignment word - padding between arrayoid and inline-data
 		 * 2. the arrayoid - an array of pointers to the leaves
 		 * 3. in-line data
-		 * In hybrid specs, the spine may also include padding for a secondary size field in empty arrays 
+		 * 4. A secondary size field, for hybrid arrays only.
```
Comment (amicic) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java]: notion of hybrid is gone, more appropriate would be discontiguous.

Comment (rwy0717) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java]: Good point, will update 👍

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java [hunk 2]
```diff
@@ -237,35 +219,27 @@ protected long getArrayLayout(J9IndexableObjectPointer array) throws CorruptData
...
 		if (GCExtensions.isVLHGC()) {
 			/* CMVC 170688:  Ensure that we don't try to allocate an inline contiguous array of a size which will overflow the region if it ever grows
 			 * (easier to handle this case in the allocator than to special-case the collectors to know how to avoid this case)
 			 * (currently, we only grow by a hashcode slot which is 4-bytes but will increase our size by the granule of alignment)
 			 */
 			minimumSpineSize = minimumSpineSize.add(ObjectModel.getObjectAlignmentInBytes());
 		}
-		
+
 		/* CMVC 135307 : when checking for InlineContiguous layout, perform subtraction as adding to dataSizeInBytes could trigger overflow. */
```
Comment (amicic) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java]: while touching this, we should probably remove the reference to CMVC

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java (hunk 1 above)
Comment (amicic) [reply debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java]: well, I got somewhat confused. concept of hybrid array layout will continue to exist. while the concept of hybrid arraylet header is about to be removed.

so to be really correct, this (4th point) should say the secondary size field exists for any hybrid and discontiguous array layout (or simpoly any non-contiguous arraylayout), right?

File: runtime/gc_base/ObjectAccessBarrier.hpp [hunk 3]
```diff
@@ -107,27 +107,20 @@ class MM_ObjectAccessBarrier : public MM_BaseVirtual
 	{
 		MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(vmThread);
 
-#if !defined(J9VM_GC_ARRAYLETS)
-		/* non-arraylet */
-		UDATA data = (UDATA)extensions->indexableObjectModel.getDataPointerForContiguous(array);
-		return (void *)(data + (elementSize * (UDATA)index));
-#else /* J9VM_GC_ARRAYLETS */
-#if defined(J9VM_GC_HYBRID_ARRAYLETS)
 		/* hybrid arraylet */
```
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.hpp]: again, this comment is referring to hybrid arraylet header, which will be on by default now. I'd just remove the comment, it creates confusion with the rest of the code block which deals with array layout (and in fact the very first one to deal with is contiguous one, not hybrid one, which the comment would suggest).

File: debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java [hunk 4]
```diff
@@ -107,16 +107,11 @@ public UDATA getSizeInElements(J9IndexableObjectPointer array) throws CorruptDat
...
 	{
 		long headerSize;
-		if (J9BuildFlags.gc_hybridArraylets) {
-			if (GC_ArrayletObjectModelBase$ArrayLayout.InlineContiguous != layout) {
-				headerSize = J9IndexableObjectDiscontiguous.SIZEOF;
-			} else {
-				headerSize = J9IndexableObjectContiguous.SIZEOF;
-			}
-		} else {
+		if (GC_ArrayletObjectModelBase$ArrayLayout.InlineContiguous != layout) {
```
Comment (dmitripivkine) [inline debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCArrayletObjectModelBase_V1.java]: Would you please reverse logical condition here?
```
if (GC_ArrayletObjectModelBase$ArrayLayout.InlineContiguous == layout) {
    headerSize = J9IndexableObjectContiguous.SIZEOF; 
} else {
    headerSize = J9IndexableObjectDiscontiguous.SIZEOF;
}
```

**Comment by dmitripivkine** [comment]
`Hybrid` can be removed from Array Layout types
```
	typedef enum ArrayLayout {
		Illegal = 0,
		InlineContiguous,
		Discontiguous,
		Hybrid
	} ArrayLayout;
```
in ArrayletObjectModelBase.hpp

**Comment by amicic** [comment]
@dmitripivkine No, it cannot, Concept of Hybrid Array Layout stays. What is being removed is the concept of (optional) hybrid arraylet header - it's always on.

File: runtime/gc_realtime/RealtimeAccessBarrier.cpp [hunk 5]
```diff
@@ -500,27 +482,10 @@ MM_RealtimeAccessBarrier::jniGetStringCritical(J9VMThread* vmThread, jstring str
...
-			isCompressed = true;
-			shouldCopy = true;
-		}
-	}
-#else /* J9VM_GC_ARRAYLETS */
-	// For now only copying is supported for arraylets
+	/* For now only copying is supported for arraylets */
 	VM_VMAccess::inlineEnterVMFromJNI(vmThread);
 	hasVMAccess = true;
 	shouldCopy = true;
```
Comment (dmitripivkine) [inline runtime/gc_realtime/RealtimeAccessBarrier.cpp]: I believe there is no need to initialize these variables to `false` earlier. We can declare variables here

File: runtime/gc_api/HeapIteratorAPI.cpp [hunk 6]
```diff
@@ -466,13 +459,6 @@ j9mm_arraylet_identification(
 	*width = 0;
 	*mask = 0;
 	*result = 0;
-#else /* J9VM_GC_ARRAYLETS */
-	*arrayletLeafSize = 0;
-	*offset = 0;
-	*width = 0;
-	*mask = 0;
-	*result = 0;
-#endif /* J9VM_GC_ARRAYLETS */
```
Comment (dmitripivkine) [inline runtime/gc_api/HeapIteratorAPI.cpp]: TODO: I can not find any caller of this function. If there is no any really it can be removed

**Comment by dmitripivkine** [comment]
@rwy0717 Would you please try to rebuild DDR? Sanity builds would not do it, so if there is a problem it would be discovered in acceptance build only.

**Comment by dmitripivkine** [comment]
Jenkins test sanity xLinux jdk11

**Pull request merged**